array/deepUnwind.ts

```diff
@@ -0,0 +1,30 @@
+import { omit } from '../object/omit';
+
+type DeepUnwind = <P extends string, T extends Record<string, any> & { [key in P]: T[] }>(array: T[], byProp: P) => Omit<T, P>[];
+
+/**
+ * Deeply unwinds array like
+ * [{ name: 'A', children: [{ name: 'A.1', children: [{ name: 'A.1.1', children: [] }] }] }]
+ * to
+ * [{ name: 'A' }, { name: 'A.1' }, { name: 'A.1.1' }]
+ */
+export const deepUnwind: DeepUnwind = (originalArray, originalByProp) => {
+  // Infinite object verification
+  try {
+    const _ = JSON.stringify(originalArray);
```

Comment:
imho ten check powinien być opcjonalny, bo ma realny wpływ na wydajność. Ktoś może mieć sytuacje, gdzie jest pewny poprawności danych, a zależy mu na wydajności. Zamiast tego można też dodać jakiś domyślny limit rekursji. To nie wpłynie na wydajność, a również zapewni bezpieczeństwo.

Reply:
@tdzienniak @tomaszwylezek  dodałem opcję, która to wyłącza.

JSON.stringify/parse zawsze kojarzyły mi się z dużą szybkością.
Ale nie zaszkodzi taki switch, do specjalistycznych zastosowań.
Natomiast tak prawdę mówiąc, jeżeli ktoś aż tak przejmuje się wydajnością, że stringify mu laguje, to w tym momencie prawdopodobnie reszta kodu tej funkcji też wymaga optymalizacji ;)

Tak czy owak, zostaje kwestia nie sprawdzania tego.

W tej chwili na commicie jest to po prostu nie sprawdzane, a test oczekuje, że się sypnie call stack exceeded. Nie widzę sensu implementować licznika, ponieważ i tak nie może być większy od rozmiaru call stacka (a ten wbrew pozorom nie jest taki duży), a przynajmniej rozmiar call stacka można zmienić.

Piszę o tym, bo zapraszam do dyskusji

Reply:
Oryginalnie w ogóle bym nie dodawał tego checka, tylko że po prostu sypało błędem o call stack exceeded, który nic by nie powiedział, co jest przyczyną.